### ISSUE TITLE
Window is not defined error when using worker

### ISSUE THREAD

**Issue by RachelRVasquez** [mholt/PapaParse#585]
Hi,

I am currently using Papa Parse with npm, which I then import into my React file using:
```
import * as Papa from 'papaparse';
```
I'm using an upload button on the front-end to pass the file to React and using Papa Parse to parse CSV files. All works fine with small files but I realized with large ones, I'd need to use a worker or else the browser crashes. However adding the worker parameter is causing an error in the console. See below for reference:
```
Papa.parse( file, {
			worker        : true,
			header        : false,
			skipEmptyLines: 'greedy',
			step          : function( results ) {
				console.log( 'Result:', results.data );
			},
			complete      : function( results ) {
				console.log('Complete', results);
			}
		} );
```

Without the worker, the error goes away - any idea how to troubleshoot this? Is it my local environment or the way Papa Parse is set up in my project? Thanks!

If it helps, the CSV file I'm testing with is the "large file" from Papa Parse's "local file" demo here: https://www.papaparse.com/demo

**Comment by pokoli** [comment]
Probably you need to set the PAPA_SCRIPT path variable as explained on the [Can Papa Parse be loaded asynchronously (after the page loads)? section of our faq](https://www.papaparse.com/faq#combine)

**Comment by RachelRVasquez** [comment]
@pokoli Cool I'll give that a try today. Thanks!

**Comment by RachelRVasquez** [comment]
@pokoli I'm still having some trouble. I have Papa Parse coming from the node_modules folder. But since I'm importing it as "Papa" in my React, I can set the PAPA_SCRIPT path using this right?
```
Papa.SCRIPT_PATH = 'papaparse/papaparse.js"
```
That didn't seem to work for me. I also tried to copy the papaparse.js in a folder with the rest of my JS instead of node_modules, and am still getting the error. Is there a way to console log whether SCRIPT_PATH is set correctly and working? If I don't set it at all, then console logging the variable itself returns null.

**Comment by RachelRVasquez** [comment]
@pokoli Do you have to use a worker when streaming? Perhaps I can just make use of the step callback? I'm not expecting to upload CSV files that have thousands of rows of data, but even testing with 100 rows seems to give the browser trouble so I'd like to stream.

**Comment by RachelRVasquez** [comment]
To answer my own question, yes I can stream using "step" or "chunk", however the browser will still freeze up. So to prevent that, the worker is necessary. 

Here is what I've tried:
- I tried setting the SCRIPT_PATH referring to it in the node modules folder, didn't work. 
- Tried setting it where it pointed to the file in a separate folder that was not in the node modules - still didn't work.
- Now I'm trying to upgrade my webpack to 4 which is another process I'm having trouble with, that has no guarantee it will resolve the issue since[ I've seen another person get the same error ](https://github.com/mholt/PapaParse/issues/566) with webpack 3.

I'm on webpack 3.12 now and still no cigar. So if workers fail because Papa parse is loaded after the page loads, then perhaps the most straightforward solution is to make sure it loads before, right?

So I loaded papa parse before my React files ever kick in. Now what happens? I get:

```
ReferenceError: Papa is not defined
```

I'm not using Typekit like that other issue I linked to - and at this point, I don't know what else to do but to look for another tool that I can implement easier with React without running into errors like these...

**Comment by RachelRVasquez** [comment]
Since I was never able to get the worker via Papa Parse working, I decided to create my own web worker instead. I found a stack overflow question with an exact example of using Papa Parse with a custom web worker:

https://stackoverflow.com/questions/38100011/using-papa-parse-for-big-csv-files/38100700#38100700

And then I modified that to my needs but within React which I needed some help with and this tutorial was what finally got it working:

https://www.fullstackreact.com/articles/introduction-to-web-workers-with-react/

So in short, if the `window is not defined` error is too much trouble to resolve due to trying to use Papa Parse's worker with your code, consider creating your own. Closing this thread. 👍

**Comment by Jordy-Guntur** [comment]
@RachelRVasquez Thank you for this. I was in the same boat - really appreciate it.

**Comment by jjech** [comment]
I don't think this issue should be closed... it appears that @RachelRVasquez, @Jordy-Guntur, and @alacret are all experiencing this issue (as am I). The FAQ as of this moment doesn't mention anything about the PAPA_SCRIPT path variable, and for using a worker it states: "Just specify worker: true in your config." I've opened #655 since I am not able to re-open this issue.

**Comment by jjech** [comment]
I'm now successfully using worker with PapaParse.

Here are the final steps that worked for me.
1. Copied papaparse.js from node_modules/papaparse/ to my main app directory.
2. Added the following lines to both App.js and index.js

var Papa = require("papaparse");
Papa.SCRIPT_PATH = "./papaparse.js";
